### PR TITLE
Adds size to profit taking logic.

### LINKED ISSUE #3452: Perpetual Market Making: profit taking order wrong size to close position
**Describe the bug**
In the event of a partial fill, the bot could end up with a profit taking order insufficient to close a position

**Steps To Reproduce**
This requires a specific sequence of order fills and exposes a flaw in the way the profit taking order is calculated and updated

A buy order for 146.7 ALGO is created
This is partially filled - 142.5 ALGO
The bot creates a profit taking order for 142.5 ALGO to close the position
The remaining portion of the initial buy order is filled
Position size is now 146.7 ALGO but profit taking order does not update. This will only occur if another buy takes place to change the entry price on the position, as the profit taking logic only considers price when determining whether to cancel existing take profit orders

**Release version**
dev 0.39.0

**Attachments**
[logs_conf_pure_mm_algo-futures.log](https://github.com/CoinAlpha/hummingbot/files/6554616/logs_conf_pure_mm_algo-futures.log)

### PR DESCRIPTION
**Pull request by petioptrv**
This PR resolves #3452.

**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in this pull request**:
Adds a size check when considering if an existing profit taking order should be cancelled and re-submitted.

**For Developers**:
- Adds size check with the same logic as the already-existing price check
- Unit tests are in progress (#3270)

**For QA**:
- Test that the profit taking order size is adjusted in the event of a partial fill of entry order

### CODE REVIEW

**Comment by dennisocana** [comment]
Thanks @petioptrv, we'll have our development team review.

**Comment by rapcmia** [comment]
Tested using Binance testnet, the bot won't create profit-taking orders until it meets the set order_amount. Upon creating initial orders, partially filled buy and the sell order got canceled. The bot then opens a position but did not create for profit-take order. Waited for more than 3mins, then filled the remaining balance and the bot created the profit-take order.

Here is one of the instances recorded for reference: 

https://user-images.githubusercontent.com/73840223/129008989-79beea57-bef0-4bb6-9974-d9cf4fc48e20.mp4

While testing, I've encountered this error but it's a hit and miss (still trying to reproduce and reported in case it happens) when I tried to do a partial fill. The bot created order since it managed to meet the order amount but getting an `unexpected error running clock-tick` with ` object has no attribute 'size'`

Test-01 - Partially filled with 5 and eventually the rmaining got filled  POS and take-profit created but getting clock-tick error

```
2021-08-11 15:51:59,002 - 11296 - hummingbot.core.clock - ERROR - Unexpected error running clock tick.
Traceback (most recent call last):
  File "hummingbot\\core\\clock.pyx", line 119, in hummingbot.core.clock.Clock.run_til
    child_iterator.c_tick(self._current_tick)
  File "hummingbot\\strategy\\perpetual_market_making\\perpetual_market_making.pyx", line 603, in hummingbot.strategy.perpetual_market_making.perpetual_market_making.PerpetualMarketMakingStrategy.c_tick
    self.c_manage_positions(session_positions)
  File "hummingbot\\strategy\\perpetual_market_making\\perpetual_market_making.pyx", line 613, in hummingbot.strategy.perpetual_market_making.perpetual_market_making.PerpetualMarketMakingStrategy.c_manage_positions
    proposals = self.c_profit_taking_feature(mode, session_positions)
  File "hummingbot\\strategy\\perpetual_market_making\\perpetual_market_making.pyx", line 661, in hummingbot.strategy.perpetual_market_making.perpetual_market_making.PerpetualMarketMakingStrategy.c_profit_taking_feature
    (o.price != price or o.size != size)
AttributeError: 'hummingbot.core.data_type.limit_order.LimitOrder' object has no attribute 'size'
```

Also, just a query about the unfilled order on partially filled instances, what will happen if the unfilled order did not get the opportunity to be filled? would it just stay hanging? thank you

File: hummingbot/strategy/perpetual_market_making/perpetual_market_making.pyx [hunk 1]
```diff
@@ -654,8 +654,15 @@ cdef class PerpetualMarketMakingStrategy(StrategyBase):
                 take_profit_price = position.entry_price * (Decimal("1") + profit_spread) if position.amount > 0 \
                     else position.entry_price * (Decimal("1") - profit_spread)
                 price = market.c_quantize_order_price(self.trading_pair, take_profit_price)
-                old_exit_orders = [o for o in active_orders if (o.price != price and position.amount < 0 and o.client_order_id in self._exit_orders and o.is_buy)
-                                   or (o.price != price and position.amount > 0 and o.client_order_id in self._exit_orders and not o.is_buy)]
+                size = position.amount
+                old_exit_orders =[
+                    o for o in active_orders
+                    if (
+                        (o.price != price or o.size != size)
```
Comment (vic-en) [inline hummingbot/strategy/perpetual_market_making/perpetual_market_making.pyx:661]:
```suggestion
                        (o.price != price or o.quantity != size)
```
Limit orders only has `quantity` property `filled_quantity` to check initial size of order and filled size respectively.
I'll suggest we don't bring in this complexity for several reasons.
I'll get back with more suggestion(s) on how we can solve the problem.

**Review by vic-en — CHANGES_REQUESTED** [review]

**Comment by PtrckM** [comment]
- after triggering stop loss order bot no longer creates new set of orders
image
image

- logs and config
[perpmm.zip](https://github.com/CoinAlpha/hummingbot/files/6986020/perpmm.zip)

**Comment by petioptrv** [comment]
> * after triggering stop loss order bot no longer creates new set of orders
>   image
>   image
> * logs and config
>   [perpmm.zip](https://github.com/CoinAlpha/hummingbot/files/6986020/perpmm.zip)

Thank you for testing @PtrckM! This bug has already a separate associated issue #3710 that I am currently working on

**Comment by petioptrv** [comment]
> Also, just a query about the unfilled order on partially filled instances, what will happen if the unfilled order did not get the opportunity to be filled? would it just stay hanging? thank you

@rapcmia just re-read the messages on this thread and saw I missed your question. Yes, the order is a profit-taker and it will remain open until filled. Thank you

**Review by RC-13 — APPROVED** [review]
Confirming that the order amount is adjusted accordingly if partial fills have taken place and the entry price is calculated correctly.

**Review by vic-en — APPROVED** [review]
LGTM

**Pull request merged**